### PR TITLE
feat: Store credentials implementation

### PR DESCRIPTION
**Pull request by pmatyjasek-sumo**
Store credentials implementation:
- Check if file with credentials exists on file system
- Read file, and use stored credentials
- Save credentials on collector shutdown

Signed-off-by: Patryk Matyjasek <pmatyjasek@sumologic.com>

### CODE REVIEW

**Comment by sumo-drosiek** [comment]
Are the credentials sensitive information?

File: pkg/extension/sumologicextension/extension.go [hunk 1]
```diff
@@ -44,6 +45,7 @@ const (
 	heartbeatUrl             = "/api/v1/collector/heartbeat"
 	registerUrl              = "/api/v1/collector/register"
 	DefaultHeartbeatInterval = 15 * time.Second
+	collectorCredentialsPath = "/tmp/sumo_otc_creds"
```
Comment (kkujawa-sumo) [inline pkg/extension/sumologicextension/extension.go:48]: Mixed block of exported and not exported variables doesn't look good to me. I suggest either splitting into two blocks or changing order  of them.

File: pkg/extension/sumologicextension/extension.go [hunk 2]
```diff
@@ -63,10 +65,19 @@ func newSumologicExtension(conf *Config, logger *zap.Logger) (*SumologicExtensio
 }
 
 func (se *SumologicExtension) Start(ctx context.Context, host component.Host) error {
-	// TODO: handle already registered collector; retrieve credentials etc.
-	if err := se.register(ctx); err != nil {
-		return err
+	var err error
+	if se.checkCollectorCredentials() {
+		se.logger.Debug("found stored credentials")
+		if err = se.getCollectorCredentials(); err != nil {
```
Comment (kkujawa-sumo) [inline pkg/extension/sumologicextension/extension.go:71]: I don't see that `err` is used later so I suggest:
```suggestion
		if err := se.getCollectorCredentials(); err != nil {
```
and removing `var err error`

File: pkg/extension/sumologicextension/extension.go [hunk 3]
```diff
@@ -63,10 +65,19 @@ func newSumologicExtension(conf *Config, logger *zap.Logger) (*SumologicExtensio
...
+			return err
+		}
+	} else {
+		se.logger.Debug("credentials not found, register collector")
+		if err = se.register(ctx); err != nil {
```
Comment (kkujawa-sumo) [inline pkg/extension/sumologicextension/extension.go:76]: the same as above

File: pkg/extension/sumologicextension/extension.go [hunk 4]
```diff
@@ -75,6 +86,10 @@ func (se *SumologicExtension) Start(ctx context.Context, host component.Host) er
 // Shutdown is invoked during service shutdown.
 func (se *SumologicExtension) Shutdown(ctx context.Context) error {
 	se.closeOnce.Do(func() { close(se.closeChan) })
+	err := se.storeCollectorCredentials()
+	if err != nil {
+		se.logger.Error("unable to store collector credentials", zap.String("error", err.Error()))
+	}
```
Comment (kkujawa-sumo) [inline pkg/extension/sumologicextension/extension.go:92]:
```suggestion
	if err := se.storeCollectorCredentials(); err != nil {
		se.logger.Error("unable to store collector credentials", zap.String("error", err.Error()))
	}
```

File: pkg/extension/sumologicextension/extension.go [hunk 5]
```diff
@@ -108,6 +123,42 @@ type OpenRegisterResponsePayload struct {
...
+	creds, err := os.Open(collectorCredentialsPath)
+	if err != nil {
+		return err
+	}
+	defer creds.Close()
+	readBytes, err := ioutil.ReadAll(creds)
+	if err != nil {
+		return err
+	}
+	json.Unmarshal(readBytes, &credentialsInfo)
```
Comment (kkujawa-sumo) [inline pkg/extension/sumologicextension/extension.go:144]: Missing error handling? https://golang.org/pkg/encoding/json/#Unmarshal

File: pkg/extension/sumologicextension/extension.go [hunk 6]
```diff
@@ -108,6 +123,42 @@ type OpenRegisterResponsePayload struct {
...
+
+func (se *SumologicExtension) storeCollectorCredentials() error {
+	data, err := json.MarshalIndent(se.registrationInfo, "", " ")
+	if err != nil {
+		return err
+	}
+	err = os.WriteFile(collectorCredentialsPath, data, 0666)
+	if err != nil {
+		return err
+	}
```
Comment (kkujawa-sumo) [inline pkg/extension/sumologicextension/extension.go:158]:
```suggestion
	if err = os.WriteFile(collectorCredentialsPath, data, 0666); err != nil {
		return err
	}
```

File: pkg/extension/sumologicextension/extension.go [hunk 7]
```diff
@@ -108,6 +123,42 @@ type OpenRegisterResponsePayload struct {
 	CollectorId            string `json:"collectorId"`
 }
 
+func (se *SumologicExtension) checkCollectorCredentials() bool {
+	if _, err := os.Stat(collectorCredentialsPath); err != nil {
```
Comment (pmalek-sumo) [inline pkg/extension/sumologicextension/extension.go:127]: I believe we should account for multiple collectors being started on particular machine hence maybe hash access keys + collector name and add it the the file name?

File: pkg/extension/sumologicextension/extension.go [hunk 8]
```diff
@@ -108,6 +123,42 @@ type OpenRegisterResponsePayload struct {
...
+
+	return nil
+}
+
+func (se *SumologicExtension) storeCollectorCredentials() error {
+	data, err := json.MarshalIndent(se.registrationInfo, "", " ")
+	if err != nil {
+		return err
+	}
+	err = os.WriteFile(collectorCredentialsPath, data, 0666)
```
Comment (pmm-sumo) [inline pkg/extension/sumologicextension/extension.go:155]: maybe 0600?

File: pkg/extension/sumologicextension/extension.go (hunk 1 above)
Comment (pmm-sumo) [inline pkg/extension/sumologicextension/extension.go:48]: I think this path is wrong. It's temporary and platform dependent. Shouldn't we device a common directory where such information is stored? Perhaps do as IC does?

Comment (pmalek-sumo) [reply pkg/extension/sumologicextension/extension.go:48]: I concur. I'm not sure what IC does but home dir makes sense. Create `~/.sumologic-otel-collector/` and store credential files there?

File: pkg/extension/sumologicextension/extension.go [hunk 9]
```diff
@@ -40,9 +46,38 @@ type SumologicExtension struct {
...
 
+type FullRegisterKeyResponse struct {
+	CollectorID   int64         `json:"collectorId"`
+	CredentialID  string        `json:"credentialId"`
+	CredentialKey string        `json:"credentialKey"`
+	CustomerID    int64         `json:"customerId"`
+	Errors        []interface{} `json:"errors"`
+	HTTPCode      int64         `json:"httpCode"`
+	Warnings      []interface{} `json:"warnings"`
+}
```
Comment (pmalek-sumo) [inline pkg/extension/sumologicextension/extension.go:57]: I was looking at linters in golangci-lint and it seems that `deadcode` doesn't detect exported types as unused :/ and [it's not configurable](https://github.com/remyoudompheng/go-misc/blob/2d6ac652a50e368c874b15d00ea5e346e240e906/deadcode/deadcode.go#L114)

Can you remove this and check other structs as well?

File: pkg/extension/sumologicextension/extension.go [hunk 10]
```diff
@@ -79,33 +122,124 @@ func (se *SumologicExtension) Shutdown(ctx context.Context) error {
...
+	se.registrationInfo = credentialsInfo
+
+	return nil
 }
 
-type OpenRegisterResponsePayload struct {
-	CollectorCredentialId  string `json:"collectorCredentialId"`
-	CollectorCredentialKey string `json:"collectorCredentialKey"`
-	CollectorId            string `json:"collectorId"`
+func (se *SumologicExtension) storeCollectorCredentials() error {
```
Comment (pmalek-sumo) [inline pkg/extension/sumologicextension/extension.go:169]: comment

File: pkg/extension/sumologicextension/extension.go [hunk 11]
```diff
@@ -79,33 +122,124 @@ func (se *SumologicExtension) Shutdown(ctx context.Context) error {
...
 }
 
-type OpenRegisterRequestPayload struct {
-	CollectorName string `json:"collectorName"`
-	Ephemeral     bool   `json:"ephemeral"`
-	Description   string `json:"description"`
-	Hostname      string `json:"hostname"`
-	Category      string `json:"category"`
-	Timezone      string `json:"timeZone"`
+func (se *SumologicExtension) getCollectorCredentials() error {
```
Comment (pmalek-sumo) [inline pkg/extension/sumologicextension/extension.go:142]: comment

File: pkg/extension/sumologicextension/extension.go [hunk 12]
```diff
@@ -79,33 +122,124 @@ func (se *SumologicExtension) Shutdown(ctx context.Context) error {
...
+			return err
+		}
+	}
+	return nil
+}
+
+// create hash key for encrypt and decrypt functions
+func createHash(key string) string {
+	hasher := md5.New()
+	hasher.Write([]byte(key))
```
Comment (pmalek-sumo) [inline pkg/extension/sumologicextension/extension.go:203]: error check, not sure why golangci-lint doesn't detect this :(

**Review by pmalek-sumo — CHANGES_REQUESTED** [review]

File: pkg/extension/sumologicextension/config.go [hunk 13]
```diff
@@ -43,6 +43,8 @@ type Config struct {
 	ApiBaseUrl string `mapstructure:"api_base_url"`
 
 	HeartBeatInterval time.Duration `mapstructure:"heartbeat_interval"`
+
+	CollectorCredentialsPath string `mapstructure:"collector_credentials_path"`
```
Comment (pmalek-sumo) [inline pkg/extension/sumologicextension/config.go:47]: Comment

File: pkg/extension/sumologicextension/factory.go [hunk 14]
```diff
@@ -38,10 +39,17 @@ func NewFactory() component.ExtensionFactory {
 }
 
 func createDefaultConfig() config.Extension {
+	homePath, err := os.UserHomeDir()
+	if err != nil {
+		return nil
+	}
+	defaultCredsPath := homePath + collectorCredentialsDirectory
```
Comment (pmalek-sumo) [inline pkg/extension/sumologicextension/factory.go:46]: Use https://golang.org/pkg/path/#Join

File: pkg/extension/sumologicextension/extension.go [hunk 15]
```diff
@@ -79,33 +112,147 @@ func (se *SumologicExtension) Shutdown(ctx context.Context) error {
...
+	}
+	nonce := make([]byte, gcm.NonceSize())
+	if _, err = io.ReadFull(rand.Reader, nonce); err != nil {
+		return nil, err
+	}
+	ciphertext := gcm.Seal(nonce, nonce, data, nil)
+	return ciphertext, nil
+}
+
+// function decrypts collector credentials stored in the file
```
Comment (pmalek-sumo) [inline pkg/extension/sumologicextension/extension.go:234]:
```suggestion
// decrypt decrypts collector credentials encrypted with AES passphrase.
```

File: pkg/extension/sumologicextension/extension.go [hunk 16]
```diff
@@ -79,33 +112,147 @@ func (se *SumologicExtension) Shutdown(ctx context.Context) error {
...
+	block, err := aes.NewCipher(key)
+	if err != nil {
+		return nil, err
+	}
+	gcm, err := cipher.NewGCM(block)
+	if err != nil {
+		return nil, err
+	}
+	nonceSize := gcm.NonceSize()
+	nonce, ciphertext := data[:nonceSize], data[nonceSize:]
```
Comment (pmalek-sumo) [inline pkg/extension/sumologicextension/extension.go:250]: Don't we need to check the `data` length for `nonceSize`?

File: pkg/extension/sumologicextension/extension.go [hunk 17]
```diff
@@ -79,33 +112,147 @@ func (se *SumologicExtension) Shutdown(ctx context.Context) error {
...
+func createHash(key string) (string, error) {
+	hasher := md5.New()
+	if _, err := hasher.Write([]byte(key)); err != nil {
+		return "", err
+	}
+	return hex.EncodeToString(hasher.Sum(nil)), nil
+}
+
+// function encrypts collector credentials to store them in the file
+func encrypt(data []byte, passphrase string) ([]byte, error) {
```
Comment (pmalek-sumo) [inline pkg/extension/sumologicextension/extension.go:213]: Ditto

```suggestion
// encrypt encrypts collector credentials with AES using the passphrase
func encrypt(data []byte, passphrase string) ([]byte, error) {
```

File: pkg/extension/sumologicextension/extension.go [hunk 18]
```diff
@@ -79,33 +112,147 @@ func (se *SumologicExtension) Shutdown(ctx context.Context) error {
...
+		err := os.Mkdir(path, 0600)
+		if err != nil {
+			return err
+		}
+	}
+	return nil
+}
+
+// create hash key for encrypt and decrypt functions
+func createHash(key string) (string, error) {
```
Comment (pmalek-sumo) [inline pkg/extension/sumologicextension/extension.go:204]: ditto

File: pkg/extension/sumologicextension/config.go [hunk 19]
```diff
@@ -36,13 +36,16 @@ type Config struct {
...
-	// CollectorCategory is the collecot rcategory which will be used when the
+	// CollectorCategory is the collecor category which will be used when the
 	// collector is being registered.
 	CollectorCategory string `mapstructure:"collector_category"`
 
 	ApiBaseUrl string `mapstructure:"api_base_url"`
 
 	HeartBeatInterval time.Duration `mapstructure:"heartbeat_interval"`
+	// CollectorCredentialsPath is path to directory where collector credentials
+	// are stored.
```
Comment (pmalek-sumo) [inline pkg/extension/sumologicextension/config.go:47]:
```suggestion
	// CollectorCredentialsPath is the path to directory where collector credentials
	// are stored.
```

Mention the defaults in the comment

File: pkg/extension/sumologicextension/config.go [hunk 20]
```diff
@@ -36,13 +36,16 @@ type Config struct {
 	// CollectorDescription is the description which will be used when the
 	// collector is being registered.
 	CollectorDescription string `mapstructure:"collector_description"`
-	// CollectorCategory is the collecot rcategory which will be used when the
+	// CollectorCategory is the collecor category which will be used when the
```
Comment (pmalek-sumo) [inline pkg/extension/sumologicextension/config.go:39]:
```suggestion
	// CollectorCategory is the collector category which will be used when the
```

File: pkg/extension/sumologicextension/extension.go [hunk 21]
```diff
@@ -79,33 +113,150 @@ func (se *SumologicExtension) Shutdown(ctx context.Context) error {
...
-type FullRegisterKeyResponse struct {
-	CollectorID   int64         `json:"collectorId"`
-	CredentialID  string        `json:"credentialId"`
-	CredentialKey string        `json:"credentialKey"`
-	CustomerID    int64         `json:"customerId"`
-	Errors        []interface{} `json:"errors"`
-	HTTPCode      int64         `json:"httpCode"`
-	Warnings      []interface{} `json:"warnings"`
+// verification if there are stored credentials for collector
+func (se *SumologicExtension) checkCollectorCredentials() bool {
```
Comment (pmalek-sumo) [inline pkg/extension/sumologicextension/extension.go:125]:
```suggestion
// checkCollectorCredentials checks if collector credentials can be found in path
// configured in the config.
func (se *SumologicExtension) checkCollectorCredentials() bool {
```

File: pkg/extension/sumologicextension/extension.go [hunk 22]
```diff
@@ -79,33 +113,150 @@ func (se *SumologicExtension) Shutdown(ctx context.Context) error {
...
 }
 
-type OpenRegisterRequestPayload struct {
-	CollectorName string `json:"collectorName"`
-	Ephemeral     bool   `json:"ephemeral"`
-	Description   string `json:"description"`
-	Hostname      string `json:"hostname"`
-	Category      string `json:"category"`
-	Timezone      string `json:"timeZone"`
+// get and decrypt collector credentials content stored in file
```
Comment (pmalek-sumo) [inline pkg/extension/sumologicextension/extension.go:138]:
```suggestion
// getCollectorCredentials retrieves and decrypts collector credentials using
// collector name as passphrase.
```

File: pkg/extension/sumologicextension/extension.go [hunk 23]
```diff
@@ -79,33 +113,150 @@ func (se *SumologicExtension) Shutdown(ctx context.Context) error {
...
+		return err
+	}
+	if err = os.WriteFile(path, encrypedCreds, 0600); err != nil {
+		return err
+	}
+	return nil
+}
+
+// check if directory to store credentials exists, if not try to create it
+func ensureStoreCredentialsDir(path string) error {
```
Comment (pmalek-sumo) [inline pkg/extension/sumologicextension/extension.go:194]:
```suggestion
// ensureStoreCredentialsDir checks if directory to store credentials exists, if not try to create it.
func ensureStoreCredentialsDir(path string) error {
```

**Review by pmalek-sumo — CHANGES_REQUESTED** [review]
Can we add a test that will actually start the extension, stop it and then check if stored credentials where retrieved successfully?

This would require an httptest server to mock the registration

File: pkg/extension/sumologicextension/config.go (hunk 13 above)
Comment (pmalek-sumo) [reply pkg/extension/sumologicextension/config.go:48]: Please mention the default in the comment here

File: pkg/extension/sumologicextension/extension_test.go [hunk 24]
```diff
@@ -93,3 +95,42 @@ func TestBasicStart(t *testing.T) {
...
+	cfg.CollectorCredentialsPath = homePath
+	se, err := newSumologicExtension(cfg, zap.NewNop())
+	require.NoError(t, err)
+	fileName, err := createHash(cfg.CollectorName)
+	require.NoError(t, err)
+	credsPath := path.Join(homePath, fileName)
+	require.NoError(t, se.Start(context.Background(), componenttest.NewNopHost()))
+	require.NoError(t, se.Shutdown(context.Background()))
+	require.FileExists(t, credsPath)
+	// To make sure that collector is usign credentials file, turn off the fake server
```
Comment (pmalek-sumo) [inline pkg/extension/sumologicextension/extension_test.go:130]:
```suggestion
	// To make sure that collector is using credentials file, turn off the mock registration server
```

File: pkg/extension/sumologicextension/extension.go [hunk 25]
```diff
@@ -40,9 +47,28 @@ type SumologicExtension struct {
...
+type OpenRegisterResponsePayload struct {
+	CollectorCredentialId  string `json:"collectorCredentialId"`
+	CollectorCredentialKey string `json:"collectorCredentialKey"`
+	CollectorId            string `json:"collectorId"`
+}
+
+const (
+	heartbeatUrl                  = "/api/v1/collector/heartbeat"
+	registerUrl                   = "/api/v1/collector/register"
+	collectorCredentialsDirectory = "/.sumologic-otel-collector/"
```
Comment (pmalek-sumo) [inline pkg/extension/sumologicextension/extension.go:68]:
```suggestion
	collectorCredentialsDirectory = ".sumologic-otel-collector/"
```

**Review by pmalek-sumo — APPROVED** [review]

**Pull request merged**